src/sqlfluff/rules/L018.py

```diff
@@ -89,66 +95,72 @@ def indent_size_up_to(segs):
                 indent_size = len(indent_str)
                 return indent_size, indent_str
 
-            balance = 0
             with_indent, with_indent_str = indent_size_up_to(raw_stack_buff)
+            # Find the end brackets for the CTE *query* (i.e. ignore optional
+            # list of CTE columns).
+            cte_end_brackets = IdentitySet()
+            for cte in context.functional.segment.children(
+                sp.is_type("common_table_expression")
+            ).iterate_segments():
+                cte_end_brackets.add(
+                    cte.children()
+                    .last(sp.is_type("bracketed"))
+                    .children()
+                    .last(sp.is_name("end_bracket"))[0]
+                )
```

Comment:
This code could've been simplified or avoided if `CTEDefinitionSegment` had **segment** definitions for the optional column list and/or the query part. I thought about it, but at least one dialect besides ANSI specializes `CTEDefinitionSegment`, so let's do it here instead. Something to keep in mind, though, if in future we have other rules with similar needs.